Move `KeyPair` to test lib

teodanciu opened the pull request:
# Description

Closes https://github.com/input-output-hk/cardano-ledger/issues/3186

# Checklist

- [x] Commit sequence broadly makes sense
- [x] Commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [x] Any changes are noted in the [changelog](../CHANGELOG.md)
- [x] Code is formatted with ormolu (which can be run with `scripts/ormolise.sh`
- [ ] Self-reviewed the diff

teodanciu committed:
Rename `KeyPair` transformation functions in a consistent way

teodanciu committed:
Remove duplicated `mkAddr` and replace usage with the `KeyPair` one

teodanciu committed:
Update Changelog to reflect moving `KeyPair`

JaredCorduan commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/API/Types.hs:
```diff
@@ -50,7 +50,7 @@ import Cardano.Ledger.Keys as X
     KESignable,
     KeyHash (..),
-    KeyPair (..),
+    -- KeyPair (..),
```
I think the right thing to do here is to make a new type, also called `KeyPair`, and deprecate it with a helpful message telling people where to find the new one.

teodanciu replied:
Makes sense!

JaredCorduan commented on libs/cardano-ledger-core/src/Cardano/Ledger/Keys.hs:
```diff
@@ -187,23 +187,23 @@ instance
+-- instance Crypto c => NoThunks (KeyPair kd c)
 
-instance HasKeyRole KeyPair
+-- instance HasKeyRole KeyPair
```
dead code ?

teodanciu replied:
eeek

JaredCorduan commented on libs/cardano-ledger-pretty/src/Cardano/Ledger/Pretty.hs:
```diff
@@ -84,12 +84,12 @@ import Cardano.Ledger.Keys
     VerKeyKES,
     hashKey,
   )
+-- import Test.Cardano.Ledger.Core.KeyPair (KeyPair (..))
```
```suggestion
```

JaredCorduan reviewed (commented):
I'm very happy to be moving this to the tests, thank you! I had just one comment about deprecating `KeyPair` from the API.

teodanciu commented on libs/cardano-ledger-core/src/Cardano/Ledger/Keys.hs:
```diff
@@ -188,23 +188,8 @@ instance
-
-instance HasKeyRole KeyPair
+newtype KeyPair c = KeyPair ()
+{-# DEPRECATED KeyPair "Use `Test.Cardano.Ledger.Core.KeyPair (KeyPair)` instead" #-}
```
@JaredCorduan  - something like this maybe?  I mean leaving it here where the original type was but deprecated, or defining it somewhere else ?

JaredCorduan replied:
yea, exactly, except that I think we want the type to remain the same (ie keep the `KeyRole` and the two fields). The idea is that we want downstream code to still compile, just emit a warning (I don't think that the warning will fire (or be as obvious) if it doesn't type check, but we could test that out).

teodanciu replied:
Yeah fair enough, at first I left it as you say but then I wanted to get rid of the imports. But you're right, if we're gonna deprecate it, it's stupid to change the type and leave the name.

JaredCorduan replied:
I am still learning too!

teodanciu committed:
Move `KeyPair` type and related functions to ledger-core testlib module

since it's only needed in tests, and deprecate the currently exported
`KeyPair` and `KeyPairs` types

teodanciu commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/LedgerState/Types.hs:
```diff
@@ -16,6 +16,7 @@
 {-# LANGUAGE TypeFamilies #-}
 {-# LANGUAGE TypeOperators #-}
 {-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -Wno-deprecations #-}
```
I checked that trying to use deprecated types from this module (eg: KeyPairs), WILL result in deprecation error, so this no-deprecation applies within this file rather than affecting the usages.

JaredCorduan replied:
awesome, good to know, thank you!

JaredCorduan reviewed (approved):
looks great, thank you!

lehins reviewed (approved):
:+1: 

Awesome. I don't think anyone outside our tests uses the `KeyPair` type, but just in case deprecating is not a bad idea.
